src/pcre2_compile_class.c

```diff
@@ -2151,40 +2182,22 @@ switch (meta)
   else
     {
     PCRE2_UCHAR flags;
-    PCRE2_UCHAR *map_start;
-    PCRE2_UCHAR *data_start;
-    PCRE2_SIZE put_length;
 
     PCRE2_ASSERT(*code_start == OP_XCLASS);
     *code_start = pop_info->op_single_type = ECL_XCLASS;
 
     PCRE2_ASSERT(code - code_start >= 1 + LINK_SIZE + 1);
     flags = code_start[1 + LINK_SIZE];
 
-    if ((flags & XCL_MAP) != 0)
+    if ((flags & XCL_MAP) != 0 ||
+        (*context->eclass_start & CLASS_ECLASS_HAS_BITMAP) != 0)
```

Comment:
I don't think you can check the CLASS_ECLASS_HAS_BITMAP flag yet.

You haven't yet scanned the entire extended class, so you don't know the final value of this. You can only check that flag at the very very end.

What you want to do is to always copy the bits from `cb->classbits` to `pop_info` (unconditionally), and always clear the XCL_MAP flag; but conditionally do `if (flags & XCL_MAP) eclass_start |= CLASS_ECLASS_HAS_BITMAP`

Reply:
It was designed with the bit is preserved in the META_ECLASS in mind. The new method requires the copy all the time as you said.